kubeadm/v1.15.0/KubeCluster.ps1

```diff
@@ -0,0 +1,273 @@
+$hnsDestination = "$PSScriptRoot\hns.psm1" 
+Invoke-WebRequest $hnsPath -o $hnsDestination
+Import-Module $hnsDestination
+
+
+ReadKubeclusterConfig -ConfigFile $ConfigFile
+InitHelper
+PrintConfig
+WriteKubeClusterConfig
+
+
+# Initialize internal network modes of windows corresponding to 
+# the plugin used in the cluster
+$Global:NetworkName = "cbr0"
+$Global:NetworkMode = "l2bridge"
+if ($Global:NetworkPlugin -eq "vxlan")
+{
+    $Global:NetworkMode = "overlay"
+    $Global:NetworkName = "vxlan0"
+}
+######################################################################################################################
+
+# Handle --install
+if ($install.IsPresent)
+{
+    InstallContainersRole
+    if (!(Test-Path $env:HOMEDRIVE/$env:HOMEPATH/.ssh/id_rsa.pub))
+    {
+        $res = Read-Host "Do you wish to generate a SSH Key & Add it to the Linux control-plane node [Y/n] - Default [Y] : "
+        if ($res -eq '' -or $res -eq 'Y'  -or $res -eq 'y')
+        {
+            ssh-keygen.exe
+        }
+    }
+
+    $pubKey = Get-Content $env:HOMEDRIVE/$env:HOMEPATH/.ssh/id_rsa.pub
+    Write-Host "Execute the below commands on the Linux control-plane node ($Global:MasterIp) to add this Windows node's public key to its authorized keys"
+    
+    Write-Host "touch ~/.ssh/authorized_keys"
+    Write-Host "echo $pubKey >> ~/.ssh/authorized_keys"
+
+    $res = Read-Host "Continue to Reboot the host [Y/n] - Default [Y] : "
+    if ($res -eq '' -or $res -eq 'Y'  -or $res -eq 'y')
+    {
+        Restart-Computer -Force
+    }
+
+    InstallCRI $Global:Cri
+    InstallKubernetesBinaries -Destination  $Global:BaseDir -Source $Global:ClusterConfiguration.Kubernetes.Source
+
+    exit
+}
+
+# Handle -Join
+if ($Join.IsPresent)
+{
+    $kubeConfig = GetKubeConfig
+    if (!(KubeConfigExists))
+    {
+        # Fetch KubeConfig from the master
+        DownloadKubeConfig -Master $Global:MasterIp -User $Global:MasterUsername
+        if (!(KubeConfigExists))
+        {
+            throw $kubeConfig + " does not exist. Cannot connect to the control-plane node"
+        }
+    }
+
+    # Validate connectivity with the API Server
+
+    Write-Host "Trying to connect to the Kubernetes control-plane node"
+    try {
+        ReadKubeClusterInfo 
+    } catch {
+        throw "Unable to connect to the control-plane node. Reason [$_]"
+    }
+
+    $KubeDnsServiceIP = GetKubeDnsServiceIp
+    $ClusterCIDR = GetClusterCidr
+    $ServiceCIDR = GetServiceCidr
+    
+    Write-Host "############################################"
+    Write-Host "Able to connect to the control-plane node"
+    Write-Host "Discovered the following"
+    Write-Host "Cluster CIDR    : $ClusterCIDR"
+    Write-Host "Service CIDR    : $ServiceCIDR"
+    Write-Host "DNS ServiceIp   : $KubeDnsServiceIP"
+    Write-Host "############################################"
+
+    #
+    # Install Services & Start in the below order
```

Comment:
After a possible reboot, there is no guarantee the services will start up again in the right order, or?
Is it possible to add dependency chain using sc after registering the services? E.g.
`sc config ServiceA depend= ServiceB`

Reply:
Kube-Proxy and Flannel don't need to be started in a particular order anymore.

@PatrickLang Does it matter what order kubelet and kubeproxy start on reboot?

Reply:
If it helps, I tried validating launching kube-proxy and kubelet in reverse order on a new WS2019 machine against an existing cluster and didn't notice any difference

Reply:
Okay, FlannelD can be started before Kubelet as well. So I think we don't need to create dependent services. However, reboot isn't officially supported yet.

Reply:
True, I should've said "crashes" instead of "reboot" -  I'll be sure to document that reboot isn't supported yet as well so that expectations are clear 👍